src/core/JobQueue.lua

```diff
@@ -0,0 +1,39 @@
+return function(max_concurrent)
+  local jobs = {}
+  local head = 1
+  local tail = 1
+  local count = 0
+  local running = 0
+
+  local execute_job
+
+  local function job_finished()
+    running = running - 1
+    if count > 0 then
+      local next = jobs[head]
+      head = head + 1
+      count = count - 1
+      execute_job(next)
+    end
+  end
+
+  execute_job = function(job)
+    coroutine.wrap(function()
+      job()
+      job_finished()
+    end)()
```

Comment:
By wrapping this in a coroutine the job essentially becomes a separately scheduled process that can be paused without holding up everything else.